### PR TITLE
interfaces/builtin: introduce dsp interface

### PR DESCRIPTION
**Pull request by anonymouse64**
This interface will be used by specific devices with plugs that have flavor
attributes identifying which will provide board specific access to digital
signal processor devices which are very specific to a particular BSP or board.

The initial accesses here are sufficient for the software that we have tested 
and have access to, we will likely need to expand this set after getting more
feedback from the customer, but in the meantime we have agreement to move
forward with merging just this for now and add things as followups.

cc @woodrow-shen

### CODE REVIEW

File: interfaces/policy/basedeclaration_test.go [hunk 1]
```diff
@@ -598,7 +598,6 @@ var (
 		"cups":                    {"app"},
 		"cups-control":            {"app", "core"},
 		"dbus":                    {"app"},
-		"docker-support":          {"core"},
```
Comment (mvo5) [inline interfaces/policy/basedeclaration_test.go:601]: Why is this removed here?

**Comment by mvo5** [comment]
Added Samuele so that the names are right.

**Review by mvo5 — APPROVED** [review]
Thanks, this looks fine to me.

File: interfaces/policy/basedeclaration_test.go (hunk 1 above)
Comment (anonymouse64) [reply interfaces/policy/basedeclaration_test.go:601]: whoops sorry this shouldn't be removed, an artifact of my editor trying to apply a formatter to this file every time I open it and needing to discard the other changes

Comment (anonymouse64) [reply interfaces/policy/basedeclaration_test.go]: removed this

File: interfaces/builtin/dsp_control.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  dsp-control:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+const dspControlBaseDeclarationSlots = `
+  dsp-control:
+    allow-installation:
+      slot-snap-type:
+        - core
```
Comment (pedronis) [inline interfaces/builtin/dsp_control.go:40]: @anonymouse64  would it make sense instead to make it a hardware interface were the slot needs to be declated on the gadget with the flavor for the device? and then the plug needs to match?

Comment (anonymouse64) [reply interfaces/builtin/dsp_control.go:40]: we could do it that way instead if you prefer, I'm not sure if it makes sense to have multiple flavor plugs on a device, it probably doesn't in which case having the slot come from the gadget makes sense

Comment (pedronis) [reply interfaces/builtin/dsp_control.go:40]: Yes, I think it would make more sense because then as I said the gadget slot would declare what flavor it exposes, given that the device nodes made accessible would/could vary by flavor

Comment (anonymouse64) [reply interfaces/builtin/dsp_control.go:40]: okay I will refactor this to be a gadget exposed slot with the flavor declared in the slot like the gpio/pwm slots. Do you think we should just make this interface something like `dsp` instead of `dsp-control` now?

Comment (pedronis) [reply interfaces/builtin/dsp_control.go:40]: yes, dsp might be a better name if this is essentially a hardware interface, though the microcode loading might be too much for the plain name

Comment (anonymouse64) [reply interfaces/builtin/dsp_control.go]: okay, I have refactored this PR (via force push) to be a gadget slotted interface with the flavor attribute on the slot side, and also removed the base declaration plug bits to match i.e. i2c, gpio, etc. and then also renamed to dsp

**Review by mvo5 — APPROVED** [review]
Thank you, looks great

File: interfaces/builtin/dsp.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+      slot-snap-type:
+        - gadget
+        - core
+    deny-auto-connection: true
+`
+
+const ambarellaDspConnectedPlugApparmor = `
+# Description: can manage and control the integrated digital signal processor on
+# the ambarella device. This allows privileged access to hardware and kernel 
+# drivers related to the digital signal processor and thus is super-privileged.
```
Comment (alexmurray) [inline interfaces/builtin/dsp.go:44]: This comment  claims the interface is super-privileged but there is nothing in the base declaration which makes it such that I can see - should there be a

```go
const dspBaseDeclarationPlugs = `
  dsp:
    allow-installation: false
    deny-auto-connection: true
`
```

**Review by alexmurray — COMMENTED** [review]
LGTM except the seeming mismatch between the comment which states this is super-privileged but then nothing which enforces this from what I can see - can this be clarified? I am not sure what purpose the hardware has and whether it makes sense for it to be super-privileged or not but it feels pretty special purpose and hence probably should be.

File: interfaces/builtin/dsp.go (hunk 3 above)
Comment (anonymouse64) [reply interfaces/builtin/dsp.go:44]: You're right, this interface was refactored to be provided by a gadget instead of snapd, and the interface is no longer super-privileged. I will adjust the comment here to reflect that. You can think of this interface as similar to gpio or pwm, etc. but technically it could be more powerful, allowing direct access to a co-processor as opposed to just hardware state of pins, but in some deployments access to the gpio or pwm pins could be considered more privileged than the DSP, so it's dependent on how it's used. Hence why this slot is now provided by a gadget.

Comment (alexmurray) [reply interfaces/builtin/dsp.go]: Thanks - +1 from me.

File: interfaces/builtin/dsp.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+func (iface *dspInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// check the flavor of the slot
+
+	var flavor string
+	_ = slot.Attr("flavor", &flavor)
+	fmt.Println("slot flavor", flavor)
+	switch flavor {
+	// only supported flavor for now
+	case "ambarella":
```
Comment (pedronis) [inline interfaces/builtin/dsp.go:81]: will there be other ambarella variants?  is this name precise enough?

**Comment by mvo5** [comment]
@woodrow-shen I think this is ready to merge - did you got a chance to test this on real hardware? Does it fit the intended purpose?

**Review by pedronis — COMMENTED** [review]
two questions

File: interfaces/builtin/dsp.go (hunk 4 above)
Comment (anonymouse64) [reply interfaces/builtin/dsp.go:81]: @woodrow-shen can you comment on this ?

File: interfaces/builtin/dsp.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/udev"
+)
+
+const dspSummary = `allows controlling digital signal processors on certain boards`
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply interfaces/builtin/dsp.go:30]: well the interface for ambarella at least allows for both controlling and using

**Comment by woodrow-shen** [comment]
> @woodrow-shen I think this is ready to merge - did you got a chance to test this on real hardware? Does it fit the intended purpose?

Yes, I did verified part of interface and it works as expected, but unfortunately the `real user` didn't response the feedback, so I will suggest going through the merge process. Thx.

File: interfaces/builtin/dsp.go (hunk 4 above)
Comment (woodrow-shen) [reply interfaces/builtin/dsp.go:81]: According to my cases, the ambarella flavor can support cv2/cv22/s6lm soc families for the current design so I think for now the `ambarella` name should be sufficient, specifically, the interface is only used for DSP controlling.
